### PR TITLE
Implements #23

### PR DESCRIPTION
**Pull request by gabrielhomsi**
See #23

### CODE REVIEW

File: src/Functional/FirstIndexOf.php [hunk 1]
```diff
@@ -37,11 +37,9 @@ function first_index_of($collection, $value)
     InvalidArgumentException::assertCollection($collection, __FUNCTION__, 1);
 
     foreach ($collection as $index => $element) {
-
-        if ($element === $value) {
+        if ($element === (is_callable($value) ? $value($element) : $value)) {
```
Comment (lstrojny) [inline src/Functional/FirstIndexOf.php]: Let’s either do a big if/else around the whole thing and implement two different loops (ugly but preferred) or assign the result of is_callable() once.

File: src/Functional/LastIndexOf.php [hunk 2]
```diff
@@ -39,7 +39,7 @@ function last_index_of($collection, $value)
     $matchingIndex = false;
 
     foreach ($collection as $index => $element) {
-        if ($element === $value) {
+        if ($element === (is_callable($value) ? $value($element) : $value)) {
             $matchingIndex = $index;
```
Comment (lstrojny) [inline src/Functional/LastIndexOf.php]: Same here

**Comment by gabrielhomsi** [comment]
Done. Furthermore, should we pass the $index variable to the callback function as a parameter? If yes, should it be the first or the second parameter (I'd rather have it as a second parameter)?

File: src/Functional/FirstIndexOf.php [hunk 3]
```diff
@@ -36,12 +36,18 @@ function first_index_of($collection, $value)
 {
     InvalidArgumentException::assertCollection($collection, __FUNCTION__, 1);
 
-    foreach ($collection as $index => $element) {
-
-        if ($element === $value) {
-            return $index;
+    if (is_callable($value)) {
+        foreach ($collection as $index => $element) {
+            if ($element === $value($element)) {
```
Comment (lstrojny) [inline src/Functional/FirstIndexOf.php]: Should be `$value($element, $index, $collection)`

File: src/Functional/LastIndexOf.php [hunk 4]
```diff
@@ -38,9 +38,17 @@ function last_index_of($collection, $value)
 
     $matchingIndex = false;
 
-    foreach ($collection as $index => $element) {
-        if ($element === $value) {
-            $matchingIndex = $index;
+    if (is_callable($value)) {
+        foreach ($collection as $index => $element) {
+            if ($element === $value($element)) {
```
Comment (lstrojny) [inline src/Functional/LastIndexOf.php]: Here as well

**Comment by gabrielhomsi** [comment]
Done!

**Comment by lstrojny** [comment]
Thank you!

**Pull request merged**